typescript/infra/src/agents/aws/validator.ts

```diff
@@ -0,0 +1,130 @@
+import { BaseValidator, types, utils } from '@abacus-network/utils';
+
+import { S3Receipt, S3Wrapper } from './s3';
+
+export enum CheckpointStatus {
+  EXTRA = '➕',
+  MISSING = '❓',
+  INVALID = '❌',
+  VALID = '✅',
+}
+
+interface CheckpointMetric {
+  status: CheckpointStatus;
+  delta?: number;
+  violation?: string;
+}
+
+// TODO: merge with types.Checkpoint
+/**
+ * Shape of a checkpoint in S3 as published by the agent.
+ */
+interface S3Checkpoint {
+  checkpoint: {
+    outbox_domain: number;
+    root: string;
+    index: number;
+  };
+  signature: {
+    r: string;
+    s: string;
+    v: number;
+  };
+}
+
+type CheckpointReceipt = S3Receipt<types.Checkpoint>;
+
+const checkpointKey = (checkpointIndex: number) =>
+  `checkpoint_${checkpointIndex}.json`;
+const LATEST_KEY = 'checkpoint_latest_index.json';
+
+/**
+ * Extension of BaseValidator that includes AWS S3 utilities.
+ */
+export class S3Validator extends BaseValidator {
+  private s3Bucket: S3Wrapper;
+
+  constructor(address: string, localDomain: number, s3Bucket: string) {
+    super(address, localDomain);
+    this.s3Bucket = new S3Wrapper(s3Bucket);
+  }
+
+  async compare(other: S3Validator): Promise<CheckpointMetric[]> {
+    const expectedLatest = await other.s3Bucket.getS3Obj<number>(LATEST_KEY);
+    const actualLatest = await this.s3Bucket.getS3Obj<number>(LATEST_KEY);
+
+    if (!expectedLatest || !actualLatest) {
+      throw new Error('Failed to get latest checkpoints');
+    }
+
+    let actualLatestIndex = actualLatest.data;
+    let expectedLatestIndex = expectedLatest.data;
+
+    const maxIndex = Math.max(actualLatestIndex, expectedLatestIndex);
+    const checkpointMetrics: CheckpointMetric[] = new Array(maxIndex + 1);
+
+    // scan extra checkpoints
+    for (; actualLatestIndex > expectedLatestIndex; actualLatestIndex--) {
+      checkpointMetrics[actualLatestIndex] = {
+        status: CheckpointStatus.EXTRA,
+      };
+    }
+
+    // scan missing checkpoints
+    for (; expectedLatestIndex > actualLatestIndex; expectedLatestIndex--) {
+      checkpointMetrics[expectedLatestIndex] = {
+        status: CheckpointStatus.MISSING,
+      };
+    }
+
+    for (; actualLatestIndex > 0; actualLatestIndex--) {
```

Comment:
This function expects to populate a status for every checkpoint, but we don't expect every checkpoint to have a status, since validators will only sign checkpoints from *after* they start running.

Reply:
yeah..., not sure how to address this other than maybe providing a date/time that the validator was started as a command line argument